test/functional/mining_mainnet.py

```diff
@@ -0,0 +1,168 @@
+
+Test mining related RPCs that involve difficulty adjustment, which
+regtest doesn't have. It uses an alternate mainnet chain.
+
+For easier testing the difficulty is maximally increased in the first (and only)
+regarget period, by producing blocks approximately 2 minutes apart.
+
+The alternate mainnet chain was generated as follows:
+- use faketime to set node clock to 2 minutes after genesis block
+- mine a block using a CPU miner such as https://github.com/pooler/cpuminer
+- restart node with a faketime 2 minutes later
+
+for i in {1..2015}
+do
+ faketime "`date -d @"$(( 1231006505 + $i * 120 ))"  +'%Y-%m-%d %H:%M:%S'`" \
+ bitcoind -connect=0 -nocheckpoints -stopatheight=$i
+done
+
+The CPU miner is kept running as follows:
+
+./minerd --coinbase-addr ... --no-stratum --algo sha256d --no-longpoll --scantime 3 --retry-pause 1
+
+This makes each block determinisic except for its timestamp and nonce, which
+are stored in data/mainnet_alt.json and used to reconstruct the chain without
+having to redo the proof-of-work.
+
+The timestamp was not kept constant because at difficulty 1 it's not sufficient
+to only grind the nonce. Grinding the extra_nonce or version field instead
+would have required additional (stratum) software. It would also make it more
+complicated to reconstruct the blocks in this test.
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+)
+from test_framework.blocktools import (
+    DIFF_1_N_BITS,
+    DIFF_1_TARGET,
+    DIFF_4_N_BITS,
+    DIFF_4_TARGET,
+    create_coinbase,
+    nbits_str,
+    target_str
+)
+
+from test_framework.messages import (
+    CBlock,
+    tx_from_hex,
+)
+
+import copy
+import json
+import os
+
+# Derived from first BIP32 test vector master key:
+# Use pkh() because tr() outputs at low heights are not spendable (unexpected-witness)
+COINBASE_OUTPUT_DESCRIPTOR="pkh(xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi/44h/0h/0h/<0;1>/*)#fkjtr0yn"
+COINBASE_SCRIPT_PUBKEY="76a914eadbac7f36c37e39361168b7aaee3cb24a25312d88ac"
+
+class MiningMainnetTest(BitcoinTestFramework):
+
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.chain = "" # main
+
+    def add_options(self, parser):
+        parser.add_argument(
+            '--datafile',
+            default='data/mainnet_alt.json',
+            help='Block data file (default: %(default)s)',
+        )
+
+        self.add_wallet_options(parser)
+
+    def mine(self, height, prev_hash, blocks, node, txs=[], fees=0):
+        self.log.debug(f"height={height}")
+        block = CBlock()
+        block.nVersion = 0x20000000
+        block.hashPrevBlock = int(prev_hash, 16)
+        block.nTime = blocks['timestamps'][height - 1]
+        block.nBits = DIFF_1_N_BITS
+        block.nNonce = blocks['nonces'][height - 1]
+        block.vtx = copy.deepcopy(txs)
```

Comment:
dec5f8ba8ff579e3e708079dde8a7ff732884479: @maflcko I'm puzzled by the need to `deepcopy` here, but if I just do `block.vtx = txs` then in the next call to `mine`, without passing any transactions in, `txs` will have the coinbase of the previous block in it.

Reply:
Well the linter forced me to change the `txs` default to `None`, which "solved" the issue.